pkg/cmd/ssh/command_test.go

```diff
@@ -0,0 +1,103 @@
+	"github.com/gardener/gardenctl-v2/pkg/config"
+	"github.com/gardener/gardenctl-v2/pkg/target"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func init() {
+	utilruntime.Must(gardencorev1beta1.AddToScheme(scheme.Scheme))
+}
+
+var _ = Describe("Command", func() {
+	It("should reject bad options", func() {
+		streams, _, _, _ := genericclioptions.NewTestIOStreams()
+		o := NewOptions(streams)
+		cmd := NewCommand(&util.FactoryImpl{}, o)
+
+		Expect(cmd.RunE(cmd, nil)).NotTo(Succeed())
+	})
+
+	It("should be able to target a garden", func() {
+		streams, _, out, _ := genericclioptions.NewTestIOStreams()
+
+		gardenName := "mygarden"
+		cfg := &config.Config{
+			Gardens: []config.Garden{{
+				Name:       gardenName,
+				Kubeconfig: "",
+			}},
+		}
+		targetProvider := internalfake.NewFakeTargetProvider(target.NewTarget("", "", "", ""))
+		factory := internalfake.NewFakeFactory(cfg, nil, nil, nil, targetProvider)
+		cmd := NewCommand(factory, NewOptions(streams))
+
+		Expect(cmd.RunE(cmd, []string{"garden", gardenName})).To(Succeed())
+		Expect(out.String()).To(ContainSubstring("Successfully targeted"))
+
+		currentTarget, err := targetProvider.Read()
+		Expect(err).NotTo(HaveOccurred())
+		Expect(currentTarget.GardenName()).To(Equal(gardenName))
+	})
+
+	It("should be able to target a project", func() {
+		streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+		gardenName := "mygarden"
+		cfg := &config.Config{
+			Gardens: []config.Garden{{
+				Name:       gardenName,
+				Kubeconfig: "",
+			}},
+		}
+
+		// user has already targeted a garden
+		currentTarget := target.NewTarget(gardenName, "", "", "")
+
+		// garden cluster contains the targeted project
+		projectName := "myproject"
+		project := &gardencorev1beta1.Project{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: projectName,
+			},
+			Spec: gardencorev1beta1.ProjectSpec{
+				Namespace: pointer.String("garden-prod1"),
+			},
+		}
+
+		fakeGardenClient := fake.NewClientBuilder().WithObjects(project).Build()
+
+		// setup command
+		targetProvider := internalfake.NewFakeTargetProvider(currentTarget)
+		clientProvider := internalfake.NewFakeClientProvider(fakeGardenClient)
+		factory := internalfake.NewFakeFactory(cfg, nil, clientProvider, nil, targetProvider)
+		cmd := NewCommand(factory, NewOptions(streams))
+
+		// run command
+		Expect(cmd.RunE(cmd, []string{"project", projectName})).To(Succeed())
+
+		currentTarget, err := targetProvider.Read()
+		Expect(err).NotTo(HaveOccurred())
+		Expect(currentTarget.GardenName()).To(Equal(gardenName))
+		Expect(currentTarget.ProjectName()).To(Equal(projectName))
+	})
+})
```

Comment:
replace target tests with proper ssh command tests ;)

Reply:
could it help if you start a local server that acts as the bastion host?

Reply:
I think at that point we firmly leave the territory of unit tests and are in full-on integration/e2e test mode.

Reply:
Let's meet in the middle and do some whitebox tests. I thought about blackbox tests, but that would require us to introduce a lot of very artifical looking interfaces to the factory that will most likely only be ever used by the SSH command.

I've added some tests, doing the minimum amount of mocking required to run most of the actual codebase.